### DISCUSSION TITLE
How can I use T-MAC with llama.cpp in Android system，,any guidence?

### DISCUSSION THREAD

**Discussion by FranzKafkaYu** [microsoft/T-MAC#12] [Q&A] [answered]
Hi developers,I am using llama.cpp in my Android Project,and I want to know how can I use T-MAC with llama.cpp.
I have tried to search info in llama.cpp but didn't find anything about T-MAC,can you give some links or guidence？

**Comment by kaleid-liner** [comment, marked as answer]
1. First, follow the normal workflow to install T-MAC (OSX/Ubuntu recommended).

2. By specifying `-u` for `run_pipeline.py`, you can use our prebuilt kernels for ARM. The performance of the kernels may not be optimal, but should be very close.

    If you want to tune and codegen kernels for your own device, you need to follow [android rpc](https://github.com/apache/tvm/blob/main/apps/android_rpc/README.md) to setup TVM RPC and [codegen doc](https://github.com/microsoft/T-MAC/blob/main/docs/codegen.md) to generate kernels.

3. For STEP.0, STEP.1, STEP.2 and STEP.3 of `run_pipeline.py`, you don't need to do any modifications.

4. For STEP.4 and STEP.5, you should follow [llama.cpp Android](https://github.com/kaleid-liner/llama.cpp/tree/6049c4cf20acc1e4c679b6ffa01b175f5cbd12d7) section to cross-compile for Android.

5. Finally, you can run STEP.6 in Android termux or adb shell.

We are also working on simplifying the deployment on Android and hopefully we can release instructions ASAP.
